Updating CONTRIBUTING.md with instructions about staging -> master set up (#313)
AlexsLemonade/refinebio-examples: CONTRIBUTING.md

_How to cherry pick commits_
Checkout your new branch.
For each commit that needs to be made live, add it to your new branch by using `git cherry-pick <commit_id>`.
Or in GitKraken, you can right click on the commit and choose `Cherry pick commit`.

_Tips for getting commits ids_  
GitKraken shows commits, but sometimes I find it hard to follow which commits belong to which branch.
If you checkout `staging` and use `git log` you can see all the most recent commits for the `staging` branch.
If you want to save it to a file for easy browsing and copy/pasting you can use this command `git log --pretty=format:'%h was %an, %ar, message: %s' > git.log`

#### Make it live, but quicker: direct merges to master, hotfixes

In (hopefully rare) scenarios where something that has already been published is noted to be broken and should be addressed quickly, [hotfix branch](https://nvie.com/posts/a-successful-git-branching-model/#hotfix-branches) pull requests are allowed.
These PRs should only be fairly small PRs and not anything that would require intense review.
This more for situations where "this is broken and here's a fix".

> After the change has been sent to `master`, you should also make sure the change still ends up in `staging`.

Author: I don't like how vague this part is, not sure how to make this less annoying but still ultimately ensure that `staging` and `master` don't get too far out of date.

Reviewer: How about something like "Once your PR  to `master` is approved, file a second PR from the branch to `staging`"

`squash and merge` will be helpful for us if we need to "cherry pick" commits if particular project or example is not yet ready for user-facing but other material is.

Previews for `staging` branch should be review using html preview.
Go here and navigate to the pertinent pages that you've changed.
`http://htmlpreview.github.io/?https://github.com/AlexsLemonade/refinebio-examples/gh-pages-stages/01-getting-started/getting-started.html`

#### Make it live: staged changes merged to master

Once a set of changes that are merged to `staging` are ready to be published and made "live", you can file a merge to `master` pull request.
These types of PRs should only involve well-polished and ready for the public material.

**Scenario 1: All changes from staging should be made live**  

If you are certain that all changes in `staging` should be carried over to `master`, first create a new branch from the most up-to-date `staging` branch.
Now you can use the new branch to create the pull request to master as you would normally do.
This allows you can resolve any merge conflicts in your new branch so we don't do a bad thing by committing conflict resolutions directly to `staging`.

> **Scenario 2: Only some changes from staging should be made live**  

Reviewer: I think both scenarios can start with a branch from `master`. In scenario 1 the next step is just to merge `staging`  into the new branch. Scenario 2 is cherry picking commits from staging.  This makes things a bit more uniform and easy to explain: "Branch from `origin/master`, then merge in what you need, which may be everything, then file a PR back to master."

Author: Okay. I like that. Will change.

docker run --mount type=bind,target=/home/rstudio,source=$PWD ccdl/refinebio-examples snakemake --cores 4
```

### Pull Requests

There are two protected branches in this repo:

`staging` - where changes are initially sent for testing and previewing.
`master` - where changes that were tested are made user-facing.

This two branch set up allows us to make incremental changes that we can test before directly making our changes "live".
It also means that pull requests have some extra methodology.

#### Add to the testing branch: merges to staging

After you've prepared the new or altered material for this repository, you can file a pull request to the `staging` branch (the default branch for this repository).
The `staging` branch does not feed into the published, user-facing material so you can feel free to make your pull requests as iterative and incremental as it is useful.

> Use `squash and merge` for these merges.

Reviewer: Add a bit more clarity about what merges we are talking about.
```suggestion
Use `squash and merge` when merging these pull requests into `staging` after approval.
```

`squash and merge` will be helpful for us if we need to "cherry pick" commits if particular project or example is not yet ready for user-facing but other material is.

> Previews for `staging` branch should be review using html preview.

Reviewer: When should this happen? 
```suggestion
After merging a pull request into `staging`, check that all the changes are correct by reviewing the newly generated pages using html preview.
```

> Go here and navigate to the pertinent pages that you've changed.
> `http://htmlpreview.github.io/?https://github.com/AlexsLemonade/refinebio-examples/gh-pages-stages/01-getting-started/getting-started.html`

Reviewer: I think it is worth adding a line about not just checking the files you expected to change, as part of the idea of staging is to make sure that everything is still in good shape after a merge.
```suggestion
Go here and navigate to the pertinent pages that you've changed.
`http://htmlpreview.github.io/?https://github.com/AlexsLemonade/refinebio-examples/gh-pages-stages/01-getting-started/getting-started.html`
You may also want to spot-check other pages to be sure that there have not been inadvertent or unexpected  changes introduced.
```

#### Make it live: staged changes merged to master

Once a set of changes that are merged to `staging` are ready to be published and made "live", you can file a merge to `master` pull request.
These types of PRs should only involve well-polished and ready for the public material.

**Scenario 1: All changes from staging should be made live**  

> - Create a new branch from the most up-to-date `master` branch.

Reviewer: Do we want a naming convention here?

Author: That's what I was asking about in my PR description but I didn't want to be giving more instructions than needed. But I think it could help streamline conversations if these kinds of branches are always named the same kind of thing. Might add some clarity. So since you've mentioned it here I think that means yes.

This strategy allows us to move forward changes from `staging` that are ready to be public facing even if other changes aren't ready (or if there isn't great timing for when the other changes will be ready).

_Tips for getting commits ids_  
GitKraken shows commits, but sometimes I find it hard to follow which commits belong to which branch.
If you checkout `staging` and use `git log` you can see all the most recent commits for the `staging` branch.
If you want to save it to a file for easy browsing and copy/pasting you can use this command `git log --pretty=format:'%h was %an, %ar, message: %s' > git.log`

#### Make it live, but quickly: direct merges to master, hotfixes

In (hopefully rare) scenarios where something that has already been published is noted to be broken and should be addressed quickly, [hotfix branch](https://nvie.com/posts/a-successful-git-branching-model/#hotfix-branches) pull requests are allowed.
These PRs should only be fairly small PRs and not anything that would require intense review.
This more for situations where "this is broken and here's a fix".

> - Create a new branch from the most up-to-date `master` branch.

Reviewer: Again, following a naming convention will be very helpful here.

- Checkout your newly created branch.
- Make the hotfix change.
- Create the pull request to `master` as you would normally do.

> - After your PR to `master` is approved, file a second PR from the same branch to `staging`.

Reviewer: Probably worth adding that the next step (or maybe even before filing the second PR) will be merging `staging` into this branch. BUT, this will cause a problem if the merge to `master` has not yet happened. So maybe this should say:
> After your PR to `master` is approved and merged...

#### A summary of types of PRs.

- `some-branch` -> `staging` -> NOT published to user-facing content.
This will be how most material is prepared so we can be more incremental with our changes and ultimately make sure only the very polished material is made user-facing.

- Group of changes in `staging` -> `master` -> published to user-facing.
This is for when the updates from the previous kinds of PRs are "ready for prime time".

> - Hotfix PR: `some-branch` -> `master` -> published to user-facing

Reviewer: ```suggestion
- Hotfix PR: `hotfix-branch` -> `master` -> published to user-facing
```

For "this-is-broken" type changes that should be hastened to the user-facing content.

> This may require a follow up merge to `staging`.

Reviewer: ```suggestion
This may require a follow up pull request and merge to `staging`.
```

After merging a pull request into `staging`, check that all the changes are correct by reviewing the newly generated pages using html preview.
Go here and navigate to the pertinent pages that you've changed.
`http://htmlpreview.github.io/?https://github.com/AlexsLemonade/refinebio-examples/gh-pages-stages/01-getting-started/getting-started.html`
You may also want to spot-check other pages to be sure that there have not been inadvertent or unexpected  changes introduced.

#### Publish: staged changes merged to master

Once a set of changes that are merged to `staging` are ready to be published, you can file a merge to `master` pull request.

These types of PRs should only involve well-polished and ready for the public material.

**Scenario 1: All changes from staging should be published**  

- Create a new branch from the most up-to-date `master` branch, call it `publish-mychanges` where `mychanges` is a short tag relevant to the changes.
- Checkout your newly created `publish` branch.
- If you are certain that all changes in `staging` should be carried over to `master` (aka published to Github pages), then you can do a `git merge staging` into your new branch.

> - Now you can use the new branch to create the pull request to master as you would normally do.

Reviewer 2: Once we merge `staging` into the new branch, will the next PR then automatically be created to merge to `master` instead of `staging`?

- Try to be as specific as possible about what PRs (and by relation, their commits) you are requesting to merge to `master`.

This strategy allows you to resolve any merge conflicts in your new branch so we don't do a bad thing by committing conflict resolutions directly to `staging`.
It also provides us with a "snapshot" if merges are continuing to happen to `staging` on other PRs -- this can make review hard if it keeps changing.

**Scenario 2: Only some changes from staging should be published**  

- Create a new branch from the most up-to-date `master` branch, call it `publish-mychanges` where `mychanges` is a short tag relevant to the changes.
- Checkout your newly created `publish` branch.
- For each commit that needs to be published, add it to your new branch by using `git cherry-pick <commit_id>`.
Or in GitKraken, you can right click on the commit and choose `Cherry pick commit`.

> - Now you can use the new branch to [create the pull request where the `master` branch](https://help.github.jp/enterprise/2.11/user/articles/creating-a-pull-request/#changing-the-branch-range-and-destination-repository) is the base branch.

Reviewer: Let's not link to the Japan pages here: 
```suggestion
- Now you can use the new branch to [create the pull request where the `master` branch](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request#changing-the-branch-range-and-destination-repository) is the base branch.
```

_Tips for getting commits ids_  
GitKraken shows commits, but sometimes I find it hard to follow which commits belong to which branch.
If you checkout `staging` and use `git log` you can see all the most recent commits for the `staging` branch.
If you want to save it to a file for easy browsing and copy/pasting you can use this command `git log --pretty=format:'%h was %an, %ar, message: %s' > git.log`

#### Publish a change, but quickly: direct merges to master, hotfixes

In (hopefully rare) scenarios where something that has already been published is noted to be broken and should be addressed quickly, [hotfix branch](https://nvie.com/posts/a-successful-git-branching-model/#hotfix-branches) pull requests are allowed.
These PRs should only be fairly small PRs and not anything that would require intense review.
This more for situations where "this is broken and here's a fix".

- Create a new branch from the most up-to-date `master` branch and call it `hotfix-mybug`, where where `mybug` is a short tag relevant to the changes.
- Checkout your newly created `hotfix` branch.
- Make the hotfix change.

> - Create the pull request to `master` as you would normally do.

Reviewer 2: As one would normally create a pull request, the default would now be to merge to `staging` right?
I guess my question here is how do we switch between merging to `master` and merging to `staging` (mainly merging to `master` since `staging` is default as of now)?

Reviewer: Good point. I think we can remove the "as you would normally do" here, as this is not "normal" for this repo, and you have to explicitly change the target, which is definitely not part of the typical process.

Author: I'll add some clarification here.

**Scenario 1: All changes from staging should be made live**  

> - Create a new branch from the most up-to-date `master` branch, call it `make-it-live`.

Reviewer: I think we should encourage specificity, something like: `make-live-mychanges`

This will also prevent errors if there are multiple branches with this name and the last one hasn't been deleted yet.

These PRs should only be fairly small PRs and not anything that would require intense review.
This more for situations where "this is broken and here's a fix".

> - Create a new branch from the most up-to-date `master` branch and call it `hotfix`.

Reviewer: Same here: `hotfix-mybug` not just `hotfix`
```suggestion
- Create a new branch from the most up-to-date `master` branch and call it `hotfix`.
```
